src/plugins/intel_cpu/src/compiled_model.cpp

```diff
@@ -102,15 +112,30 @@ CompiledModel::CompiledModel(const std::shared_ptr<ov::Model>& model,
     } else {
         CompiledModel::get_graph();
     }
-    // init sub stream threads of executor
-    int sub_streams = m_cfg.streamExecutorConfig.get_sub_streams();
-    if (sub_streams > 0 && stream_executor != nullptr) {
-        std::vector<Task> tasks;
-        tasks.resize(sub_streams);
-        for (auto&& task : tasks) {
-            task = [] {};
+    if (m_cfg.numSubStreams > 0) {
+        m_has_sub_compiled_models = true;
+        auto sub_cfg = m_cfg;
+        sub_cfg.numSubStreams = 0;
+        sub_cfg.enableNodeSplit = true;
+        auto streams_info_table = m_cfg.streamExecutorConfig.get_streams_info_table();
+        auto message = message_manager();
+        m_sub_memory_manager = std::make_shared<SubMemoryManager>(m_cfg.numSubStreams);
+        message->set_num_sub_streams(m_cfg.numSubStreams);
+        for (int i = 0; i < m_cfg.numSubStreams; i++) {
+            std::vector<std::vector<int>> sub_streams_table;
+            sub_streams_table.push_back(streams_info_table[i + 1]);
+            sub_streams_table[0][NUMBER_OF_STREAMS] = 1;
+            sub_cfg.streamExecutorConfig = IStreamsExecutor::Config{"CPUStreamsExecutor",
+                                                                    1,
+                                                                    1,
+                                                                    ov::hint::SchedulingCoreType::ANY_CORE,
+                                                                    false,
+                                                                    true,
+                                                                    sub_streams_table,
+                                                                    sub_cfg.streamsRankTable[i]};
+            m_sub_compiled_models.push_back(
+                std::make_shared<CompiledModel>(model, plugin, sub_cfg, loaded_from_cache, m_sub_memory_manager));
```

Comment:
There is a general comment regarding streams configuration.
First, we introduced substreams, which are not exactly the same as usual streams but have the same goal - execute something within a stream.
Now we have those internal streams, which are much closer to the 'normal' streams (at least they are created using the same interface), but still, for some reason, created separately, not in scope of the commit stream creation logic / routine.
I think we need (maybe even 'must') to generalize the streams creation for all the use cases we have.
So basically, before the tensor parallel, we had two main scenarios:
1. Single stream, single cpu graph, this cpu graph has no asynchronous execution inside. Only one stream is available for the user and its infer_requests.
   We call this a 'latency' mode
2. Multiple streams, each using its own cpu graph. Those cpu graphs have no asynchronous execution inside (all the nodes are executed subsequently). All the streams are available for the user and its infer_requests.
   We call this a 'throughput' mode

Basically, we need to extend this configuration and those abstractions to support two new use cases:
1. Multiple streams, multiple cpu graphs, this cpu graph has no asynchronous execution inside. Only one stream is available for the user and its infer_requests, and the rest are available only for the internal plugin logic.
   This is the tensor parallel solution implemented in scope of your PR
2. Multiple streams, single cpu graphs, this cpu graph potentially have asynchronous execution inside. Only one stream is available for the user and its infer_requests, and the rest are available only for the internal plugin logic. CPU graph have an access to all the avaliable streams.
   This is an alternative tensor parallel solution, where everything is implemented as part of cpu graph.
   
We can formalize this as a following table:

| Mode             | Total number of streams | Streams avaiable for the user | Number of cpu graphs | cpu graph streams |
|------------------|-------------------------|-------------------------------|----------------------|-------------------|
| Latency w/o TP   | 1                       | 1                             | 1                    | 0                 |
| Throughput       | N                       | N                             | N                    | 0                 |
| Latency TP       | N                       | 1                             | N                    | 0                 |
| Latency TP (alt) | N                       | 1                             | 1                    | N                 |
| Troughput TP     | M * K                   | M                             | M                    | K                 |

The last entry 'thoughput tensor parallel' is a theoretical configration we might or might not need to implement in the future. I have mentioned it just to highlight how 'flexible' the configuration should be

Reply:
Hi @EgorDuplensky, the new design of this PR introduce the stream wrapper for user and sub stream for model execution. Stream wrapper has same interface as normal stream, but without TBB arena for execution. Sub streams are totally same as normal stream. 

Then your table can be updated as below:
| Mode             | Total number of execution streams | Streams available for the user | Number of cpu graphs | Number of sub stream in each user stream |
|------------------|-------------------------|-------------------------------|----------------------|-------------------|
| Latency w/o TP   | 1                       | 1                             | 1                    | 0                 |
| Throughput       | N                       | N                             | N                    | 0                 |
| Latency TP       | N                       | 1                             | N                    | N                |
| Latency TP (alt) | N                       | 1                             | 1                    | N                 |
| Troughput TP     | M * K                   | M                             | M                    | K                 |

During model compilation, stream_info_table is created based on mode information, which include stream wrapper and sub streams information. Then StreamsExecutor will create stream wrapper and sub streams based on stream_info_table.

This PR is the implementation of Latency TP mode. So the change in stream_info_table and  StreamsExecutor only cover Latency TP mode as well. When we create PR for Latency TP (alt), we can extend stream_info_table and  StreamsExecutor to support it.

Reply:
Sure, I didn't mean that we need to implement this in scope of the current PR, since we don't really have time to do this. I just wanted to highlight the idea.